File: app/src/main/java/com/waveneuro/domain/usecase/protocol/GetLatestProtocolUseCase.java
package com.waveneuro.domain.usecase.protocol;

import com.asif.abase.domain.base.ObservableUseCase;
import com.asif.abase.domain.base.UseCaseCallback;
import com.waveneuro.data.DataManager;
import com.waveneuro.data.model.response.protocol.ProtocolResponse;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Observable;

public class GetLatestProtocolUseCase extends ObservableUseCase<ProtocolResponse> {

    private final DataManager dataManager;
    private int id;


    @Inject
    public GetLatestProtocolUseCase(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    @Override
    public Observable<ProtocolResponse> buildUseCaseSingle() {
        return dataManager.protocol(id);
    }

    public void execute(int id, UseCaseCallback<ProtocolResponse> useCaseCallback) {
        this.id=id;
        super.execute(useCaseCallback);
    }

}

File: app/src/main/java/com/waveneuro/domain/usecase/protocol/GetLatestProtocolUseCase.kt
package com.waveneuro.domain.usecase.protocol

import com.waveneuro.data.api.user.UserApi
import com.waveneuro.domain.model.protocol.ProtocolMapperImpl
import com.waveneuro.domain.model.protocol.ProtocolRs
import javax.inject.Inject

class GetLatestProtocolUseCase @Inject constructor(
    private val userApi: UserApi,
    private val mapper: ProtocolMapperImpl
) {

    suspend fun getProtocol(id: Int): ProtocolRs =
        mapper.fromApiToDomain(userApi.getProtocolForUser(id))

}